### ISSUE TITLE
When rack-timeout fires, any reason to return a 5xx and not a 408 request timeout?

### ISSUE THREAD

**Issue by bf4** [sharpstone/rack-timeout#166]
I don't see any reference to a [408](http://httpstatus.es/408) so I wanted to check if others have considered returning a non 5xx.

On the other handle, 4xx usually means the client did something wrong. In this case, I might argue that the server did something wrong by not being able to meet its promise to return quickly, and so it *should* be a 5xx.  

But I wanted to communicate to the client that this error is a timeout that *may* work if they try again as opposed to [504](https://httpstatuses.com/504) which usually feels more nefarious

related 
- https://github.com/sharpstone/rack-timeout/issues/3
- https://github.com/sharpstone/rack-timeout/issues/40

**Comment by bf4** [comment]
@dentarg thanks for the feedback.  Given that the default behavior of the gem is to return a 500, I think it's fair to recommend some other options, assuming this gem as an 'expert in what to do when it is used' :)

**Comment by dentarg** [comment]
> Given that the default behavior of the gem is to return a 500

No, this is not what this gem does, if this happens, it happens in your application (or your application framework)

This gem raise an exception when the timeout occurs, see https://github.com/sharpstone/rack-timeout/blob/v0.6.0/doc/exceptions.md

**Comment by bf4** [comment]
I understand that.  I think that's a little disingenuous to say it raises an unhandled exception with no recommendation as to how you may want to handle it.

For example, it could have a railtie which adds a default response status

```ruby
Rails.configuration.action_dispatch.rescue_responses["Rack::Timeout::Error"] = :service_unavailable
Rails.configuration.action_dispatch.rescue_responses["Rack::Timeout::RequestTimeoutException"] = :service_unavailable
```

**Comment by bf4** [comment]
Oh cute, the slowpoke gem which extends rack-timeout does just this

https://github.com/ankane/slowpoke/blob/cb7a678b0dcc0ebe906f2eeec6915620f612df95/lib/slowpoke/railtie.rb#L1-L8

```ruby
module Slowpoke
  class Railtie < Rails::Railtie
    config.slowpoke = ActiveSupport::OrderedOptions.new

    # must happen outside initializer (so it runs earlier)
    config.action_dispatch.rescue_responses.merge!(
      "Rack::Timeout::RequestTimeoutError" => :service_unavailable,
      "Rack::Timeout::RequestExpiryError" => :service_unavailable

```

I'm happy to make a PR to reference this or just close this issue

**Comment by schneems** [comment]
LMK if there's a doc PR that needs to be made. FWIW I would like to merge the examples in this repo in here https://github.com/schneems/rack_timeout_demos or at least link to them from the docs here.

**Comment by bf4** [comment]
I'm interested in helping improve usage and docs.  But need buy-in.

Maybe the issue title is the problem. I don't know.   Documenting configuring generated response code seemed a no brainer to me but this is closed now so shrug

**Comment by schneems** [comment]
I took a closer look at this issue. It looks like rack-timeout doesn't really handle status codes at all it just fires an exception and then counts on someone else to grab it and turn it into a 500/etc. Which is what @dentarg said, and I just glossed over it.

There's a doc folder: https://github.com/sharpstone/rack-timeout/tree/master/doc we could start documenting behavior in a file there. There's already https://github.com/sharpstone/rack-timeout/blob/v0.6.0/doc/exceptions.md which we could extend.

I think the options going forward would be to either add docs on how to set this up as listed in this issue or picking an appropriate default response and adding a railtie so we set to a more descriptive status code.

The only thing I don't love about `408` is the mantra "5xx is my problem 4xx is your problem". The timeout might be due to a poorly configured server and not within the user's power to address. I think the description is pretty spot on though https://www.lifewire.com/408-request-timeout-2622937. Let's re-open this since there seems to be interest.

**Comment by bf4** [comment]
Yeah,  408 is wrong,  though I believe I came to this gem to move timeouts from heroku router 499 into the app

**Comment by wuputah** [comment]
I generally put this in the category of "user can configure their app/framework as they see fit." I'm not sure if adding something to the Railtie to configure Rails to return a 503 (Service Unavailable) instead of a 500 (for an unhandled exception) provides a significant benefit. I think a 503 implies more of manual downtime rather than an issue specific to a particular request, but that's certainly subject to interpretation.

I wouldn't return a 408 -- its use here that doesn't align with this documentation.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/408
